llvm/lib/Analysis/ScalarEvolution.cpp

```diff
@@ -10137,8 +10140,17 @@ static const SCEV *SolveLinEquationWithOverflow(const APInt &A, const SCEV *B,
   //
   // B is divisible by D if and only if the multiplicity of prime factor 2 for B
   // is not less than multiplicity of this prime factor for D.
-  if (SE.getMinTrailingZeros(B) < Mult2)
-    return SE.getCouldNotCompute();
+  if (SE.getMinTrailingZeros(B) < Mult2) {
```

Comment:
Should we maybe try to prove that (urem B, A) == 0 before resorting to the predicate?

Reply:
IIUC that check shouldn't be true, otherwise we are missing some logic in `getMinTrailingZeros`? Added an assert that the check isn't true.

Reply:
With the reframing of urem (B, 1 << Mult2) == 0, I agree with your comment, but I think we could reasonable see divergence with the original urem(B,A) check.  Those are just different enough.

My suggestion is basically, can we prove the original fact without using the trailing bits proof strategy?  Said differently. what if D = gcd(A, N) is something like 3^M?

Hm, though now I see that the multiplicative inverse code below would need updated too.  I'd missed that originally.

(Totally fine to move forward here, this is purely a possible followup.)

Reply:
I stand corrected, the assertion actually uncovered a case where getMinTrailingZeros returns a worse result and I think catching this in getMinTrailingZeros would require matching a specific pattern (https://alive2.llvm.org/ce/z/t3A5X2)

I updated the code to check if the URem is zero if the trailing bits check failed, as we need to build the expression there already. Test is in `llvm/test/Analysis/ScalarEvolution/trip-count-urem.ll`